subnet/kube/kube.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	var cniConf, netConf []byte
+	cniConf, err = ioutil.ReadFile(cniConfPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cni conf: %v", err)
+	}
+	netConf, err = ioutil.ReadFile(netConfPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read net conf: %v", err)
+	}
+
+	err = ioutil.WriteFile(cniConfTargetPath, []byte(cniConf), 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write cni conf: %v", err)
+	}
+
+	sc, err := subnet.ParseConfig(string(netConf))
+	if err != nil {
+		return nil, fmt.Errorf("error parsing subnet config: %s", err)
+	}
+	sm, err := newKubeSubnetManager(c, sc, nodeName)
+	if err != nil {
+		return nil, fmt.Errorf("error creating network manager: %s", err)
+	}
+	go sm.Run(context.Background())
+	return sm, err
+}
+
+func newKubeSubnetManager(c clientset.Interface, sc *subnet.Config, nodeName string) (*kubeSubnetManager, error) {
+	var ksm kubeSubnetManager
+	ksm.client = c
+	ksm.nodeName = nodeName
+	ksm.subnetConf = sc
+	ksm.nodeStore.Store, ksm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return ksm.client.Core().Nodes().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return ksm.client.Core().Nodes().Watch(options)
+			},
+		},
+		&api.Node{},
+		resyncPeriod,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	return &ksm, nil
+}
+
+func (ksm *kubeSubnetManager) GetNetworkConfig(ctx context.Context, network string) (*subnet.Config, error) {
+	glog.V(4).Infof("get network config: %#v", kubeSubnetCfg)
+	time.Sleep(time.Second)
+	return ksm.subnetConf, nil
+}
+
+func (ksm *kubeSubnetManager) AcquireLease(ctx context.Context, network string, attrs *subnet.LeaseAttrs) (*subnet.Lease, error) {
+	glog.V(4).Infof("acquire lease")
+	time.Sleep(time.Second)
+	nobj, found, err := ksm.nodeStore.Store.GetByKey(ksm.nodeName)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, fmt.Errorf("node %q not found", ksm.nodeName)
+	}
+	n, ok := nobj.(*api.Node)
+	if !ok {
+		return nil, fmt.Errorf("nobj was not a *api.Node")
+	}
+	if n.Spec.PodCIDR == "" {
+		return nil, fmt.Errorf("node %q pod cidr not assigned", ksm.nodeName)
+	}
+	bd, err := attrs.BackendData.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	if n.Annotations[backendDataAnnotation] != string(bd) ||
+		n.Annotations[backendTypeAnnotation] != attrs.BackendType ||
+		n.Annotations[backendPublicIPAnnotation] != attrs.PublicIP.String() ||
+		n.Annotations[subnetKubeManagedAnnotation] != "true" {
+		n.Annotations[backendTypeAnnotation] = attrs.BackendType
+		n.Annotations[backendDataAnnotation] = string(bd)
+		n.Annotations[backendPublicIPAnnotation] = attrs.PublicIP.String()
+		n.Annotations[subnetKubeManagedAnnotation] = "true"
+		n, err = ksm.client.Core().Nodes().Update(n)
+		if err != nil {
+			return nil, err
+		}
+	}
+	_, cidr, err := net.ParseCIDR(n.Spec.PodCIDR)
```

Comment:
Unlikely error, but should we verify this prior to updating the node object annotations? If my understanding is correct, once the annotations are updated it is effectively a successful lease from the perspective of other nodes.